Do not allow version in Rest Update API

PnPie opened the pull request:
The versioning of Update API doesn't rely on version number anymore (and rather on sequence number). But in rest api level we ignored the "version" and "version_type" parameter, so that the server cannot raise the exception when whey were set.

This PR restores "version" and "version_type" parsing in Update Rest API so that we can get the appropriate errors.

Relates to #42497

PnPie committed:
Fix versioning for Update Rest API

The versioning of Update API doesn't rely on version number anymore (and rather on sequence number now). But in rest api level we ignored and didn't parse the "version" and "version_type" parameter, so that the server doesn't raise the exception even if we set them.

This PR restores "version" and "version_type" parsing in Update Rest API so that we can get the appropriate errors.

elasticmachine commented:
Pinging @elastic/es-distributed

dnhatn commented:
Thank you for working on this PR. I think we should not call `UpdateRequest#version()` and `UpdateRequest#versionType()` when parsing an update request because these methods always throw an exception. This would make every update request fail (you can run ` ./gradlew :distribution:archives:integ-test-zip:check`). We can instead throw an exception only if either version or version_type parameter is provided. We might also need to make the behavior of update and delete consistent in a follow-up.

PnPie committed:
Make version parsing conditional

PnPie commented:
Hi @dnhatn I see the problem and I updated it to parse `version/version_type` if provided.

dnhatn commented on server/src/main/java/org/elasticsearch/rest/action/document/RestUpdateAction.java:
```diff
@@ -83,6 +83,12 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
 
         updateRequest.retryOnConflict(request.paramAsInt("retry_on_conflict", updateRequest.retryOnConflict()));
+        if (request.hasParam("version")) {
+            updateRequest.version(RestActions.parseVersion(request));
```
We should throw ActionRequestValidationException (instead of UnsupportedOperationException) to return 400. How about this?

```java
if (request.hasParam("version") || request.hasParam("version_type")) {
	final ActionRequestValidationException versioningError = new ActionRequestValidationException();
	versioningError.addValidationError("internal versioning can not be used for optimistic concurrency control. " +
		"Please use `if_seq_no` and `if_primary_term` instead");
	throw versioningError;
}
```

PnPie replied:
I updated it and I suppose it is to be aligned with `DeleteRequest` ? Maybe we should move it to the `validate()` later and make the `version` part codes consistent with the `DeleteRequest` as you mentioned before ?

dnhatn reviewed (commented):
@PnPie I am sorry for being slow on this. I left a comment on the validation. Can you also add a YAML test ensuring that we return `bad_request` for an update request if version or version_type is provided? Thank you!

PnPie committed:
Change exception thrown for update version

Change exception thrown from UnsupportedOperationException to ActionRequestValidationException

PnPie committed:
Merge branch 'master' into update_api_version

PnPie commented:
Hi @dnhatn sorry for being slow on this too. I added a YAML test for this but it throws an exception directly [at the YAML test level](https://github.com/elastic/elasticsearch/blob/master/test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ClientYamlTestClient.java#L122-L123), before going into rest request parsing part. I see it's because in [YAML test update api spec](https://github.com/elastic/elasticsearch/blob/master/rest-api-spec/src/main/resources/rest-api-spec/api/update.json) we have removed `version/version_type` from supported Params. So that means we cannot test this Exception at YAML test layer ? By the way we can test it in `RestUpdateActionTests` and I updated it.

dnhatn commented:
@elasticmachine test this please

dnhatn commented on server/src/test/java/org/elasticsearch/rest/action/document/RestUpdateActionTests.java:
```diff
@@ -47,4 +59,44 @@ public void testTypeInPath() {
     }
+
+    public void testUpdateDocVersion() {
+        Map<String, String> params = Collections.singletonMap("version", "100");
```
Maybe combine the new tests into a single test where the parameters include either version or version_type or both.

PnPie replied:
Yes, I updated the tests.

dnhatn replied:
Sorry, I wasn't clear. I meant a single block of test with a randomized parameter. Something like this:

```java
    public void testUpdateDocVersion() {
        Map<String, String> params = new HashMap<>();
        if (randomBoolean()) {
            params.put("version", Long.toString(randomNonNegativeLong()));
            params.put("version_type", randomFrom(VersionType.values()).name());
        } else if (randomBoolean()) {
            params.put("version", Long.toString(randomNonNegativeLong()));
        } else {
            params.put("version_type", randomFrom(VersionType.values()).name());
        }
        String content =
            "{\n" +
                "    \"doc\" : {\n" +
                "        \"name\" : \"new_name\"\n" +
                "    }\n" +
                "}";
        FakeRestRequest updateRequest = new FakeRestRequest.Builder(xContentRegistry())
            .withMethod(RestRequest.Method.POST)
            .withPath("test/_update/1")
            .withParams(params)
            .withContent(new BytesArray(content), XContentType.JSON)
            .build();
        ActionRequestValidationException e = expectThrows(ActionRequestValidationException.class,
            () -> action.prepareRequest(updateRequest, mock(NodeClient.class)));
        assertThat(e.getMessage(), containsString("internal versioning can not be used for optimistic concurrency control. " +
            "Please use `if_seq_no` and `if_primary_term` instead"));
    }
```

PnPie replied:
I see it. Thanks @dnhatn

dnhatn reviewed (approved):
I left an optional comment but LGTM. Thanks @PnPie for working on this :).

PnPie committed:
Update tests

PnPie committed:
update tests

dnhatn commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into update_api_version

dnhatn commented:
@elasticmachine test this please

dnhatn commented:
@PnPie Thanks again for working on this.
